Scripts/Runtime/Data/VirtualData/InstanceData/IKeyedData.cs

```diff
@@ -1,21 +1,20 @@
-using System;
-
 namespace Anvil.Unity.DOTS.Data
 {
     /// <summary>
     /// Data to be used in <see cref="VirtualData{TKey,TInstance}"/>
     /// that can be looked up with a <typeparamref name="TKey"/>
     /// </summary>
     /// <typeparam name="TKey">The type of key to use for lookup.</typeparam>
-    public interface IKeyedData<out TKey>
-        where TKey : unmanaged, IEquatable<TKey>
+    public interface IKeyedData
     {
         /// <summary>
         /// The lookup key for this data
         /// </summary>
-        public TKey Key
+        public VDContextID ContextID
         {
             get;
+            //TODO: DISCUSS - Way to hide this set. Might not be possible.
```

Comment:
I don't think there is a way to do this, but I'll struggled long enough that it's worth talking about. 

I don't want you to be able to set the `ContextID` from outside. Only internally when I'm doing something rewrite the context part out from underneath you as the data flows across different TaskDrivers. 

I can't make it an `internal set` because then whoever implements this interface in the game code is no longer internal and can't compile.

I've tried some pointer magic to get the memory for the ContextID and then reach in and change the context but the problem is that because it's a getter, it returns a copy of the backing field. My changes work but only on that copy, the original backing field is still the same. 

We could get around that with a ref return, but of course structs aren't allowed to do that. 

So our options are this:

1. Add a set. You shouldn't use it, you'll probably break things if you do and there's no way we can guard against it. However similarly `Entity` lets you modify the `Index` and `Version` fields and you really shouldn't ever do that either.

2. Make the entire game assembly allow unsafe code. Have this interface have a function that returns a pointer to the backing field. You the developer would have to copy paste that code to return the proper pointer into every VirtualData instance struct you ever create... I think it's gross. But it does work.

Reply:
In your DotFiddle you are returning a ref to a struct from a class. 
I need to return a ref to a struct from a struct. https://dotnetfiddle.net/TYUo3D


The core issue is that you as the Developer need to create structs to represent your data. This data can be and will be anything `TimerInstance`, `WanderInstance`, `PickupTarget` etc etc etc. 

We need to store that data in a lookup and we need to associate it with an `Entity` and we need to know which `TaskDriver` it came from. 

If it was a class, no problem it all just inherits from `AbstractData` and off we go. But since it's a struct best we can do is make you implement an interface to return an inner struct for an ID. In this case a `VDContextID`. 

Because it's an interface, it must be a method/getter. Which means it always returns a copy which makes it hard to inject a new context. 

I can't make any of the data internal because you as the dev need to actually implement. 

The constructor that takes the context is internal but I need one public one so that you can seed it with your Entity. 

The only way I can lock it down nicely is to make the developer implement the VDContextID as a field on the struct and then also implement a method that returns a pointer to that field. There's no way to enforce that they did it correctly, so I just have to hope they did it right for every single data structure they create. It's a lot of boilerplate.

Reply:
Ah I see,
Discussed on phone about maybe just composing the VDContextID on demand internally.

Reply:
I reworked this so that you as the developer deal only with your instances of `IKeyedData` which operate on `Entity` as the key. 

It's readonly and can't be changed so that's good.

Internally, `VirtualData` handles wrapping your instance in a `VDWrapper` struct with assigns the proper context for you. You never see it and it's internal so you can't. 

I profiled and the difference was negligible.